internal/providers/nscloud/wait.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2022 Namespace Labs Inc; All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+
+package nscloud
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	k8s "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"namespacelabs.dev/foundation/internal/console"
+	"namespacelabs.dev/foundation/internal/executor"
+	"namespacelabs.dev/foundation/internal/fnerrors"
+	"namespacelabs.dev/foundation/internal/providers/nscloud/api"
+	"namespacelabs.dev/foundation/internal/runtime/kubernetes/client"
+	"namespacelabs.dev/foundation/internal/runtime/kubernetes/kubeobserver"
+	"namespacelabs.dev/foundation/std/tasks"
+)
+
+const (
+	kubeSystem       = "kube-system"
+	coredns          = "coredns"
+	localProvisioner = "local-path-provisioner"
+)
+
+func WaitKubeSystem(ctx context.Context, cluster *api.KubernetesCluster) error {
+	return tasks.Action("cluster.wait-kube-system").
+		Arg("id", cluster.ClusterId).Run(ctx, func(ctx context.Context) error {
+		cfg := clientcmd.NewDefaultClientConfig(MakeConfig(cluster), nil)
+		restcfg, err := cfg.ClientConfig()
+		if err != nil {
+			return fnerrors.New("failed to load kubernetes configuration: %w", err)
+		}
+
+		cli, err := k8s.NewForConfig(restcfg)
+		if err != nil {
+			return fnerrors.New("failed to create kubernetes client: %w", err)
+		}
+
+		var list *appsv1.DeploymentList
+		// First wait until at least coredns and local-path-provisioner deployments exist.
+		if err := client.PollImmediateWithContext(ctx, 500*time.Millisecond, 5*time.Minute, func(ctx context.Context) (done bool, err error) {
+			list, err = cli.AppsV1().Deployments(kubeSystem).List(ctx, metav1.ListOptions{})
+			if err != nil {
+				return false, err
+			}
+
+			names := map[string]struct{}{}
+			for _, d := range list.Items {
+				names[d.Name] = struct{}{}
+			}
+
+			if _, ok := names[coredns]; !ok {
+				return false, nil
+			}
+
+			if _, ok := names[localProvisioner]; !ok {
+				return false, nil
+			}
+
+			return true, nil
+		}); err != nil {
+			return fnerrors.New("failed to wait for deployments in namespace %q: %w", kubeSystem, err)
+		}
+
+		fmt.Fprintf(console.Debug(ctx), "got %d kube-system deployments\n", len(list.Items))
+
+		eg := executor.New(ctx, "wait")
+
+		for _, d := range list.Items {
```

Comment:
This waits for everything in kube-system, doesn't seem quite right. Should we not wait only for the deployments listed above?

Reply:
It felt more conservative to me to wait for all of kube-system.
In practice, I've only seen it wait for the two. Updated as you believe they suffice :)